lldb/include/lldb/Target/ThreadPlanStack.h

```diff
@@ -128,15 +133,34 @@ class ThreadPlanStackMap {
 
   void AddThread(Thread &thread) {
     lldb::tid_t tid = thread.GetID();
-    m_plans_list.emplace(tid, thread);
+    // If we already have a ThreadPlanStack for this thread, use it.
+    if (m_plans_list.find(tid) != m_plans_list.end())
+      return;
+
+    m_plans_up_container.emplace_back(
+        std::make_unique<ThreadPlanStack>(thread));
+    m_plans_list.emplace(tid, m_plans_up_container.back().get());
   }
 
   bool RemoveTID(lldb::tid_t tid) {
     auto result = m_plans_list.find(tid);
     if (result == m_plans_list.end())
       return false;
-    result->second.ThreadDestroyed(nullptr);
+    ThreadPlanStack *removed_stack = result->second;
     m_plans_list.erase(result);
+    // Now find it in the stack storage:
+    PlansStore::iterator end = m_plans_up_container.end();
+    PlansStore::iterator iter = std::find_if(m_plans_up_container.begin(), end,
+        [&] (std::unique_ptr<ThreadPlanStack> &stack) {
+          return stack->IsTID(tid);
+        });
+    if (iter == end)
+      return false;
+
+    // Then tell the stack its thread has been destroyed:
+    removed_stack->ThreadDestroyed(nullptr);
+    // And then remove it from the container so it goes away.
+    m_plans_up_container.erase(iter);
```

Comment:
FYI: If haven't discovered it yet, there is also std::remove_if (https://en.cppreference.com/w/cpp/algorithm/remove). It works by moving the to-be-removed element to the end so you can std::erase it there (see the example).

Reply:
Thanks, I'll keep that in mind.  ! I don't think that would make this code any clearer.